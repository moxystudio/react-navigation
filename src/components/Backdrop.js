import React, { useCallback, useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const Backdrop = ({ className, isDrawerOpen, toggleDrawer }) => {
    const backdropClassName = useMemo(() => classNames(
        'react-navigation_backdrop',
        { show: isDrawerOpen },
        className,
    ), [className, isDrawerOpen]);

    const handleKeyDown = useCallback((event) => {
        const ESC_KEYCODE = 27;

        if (isDrawerOpen && event.keyCode === ESC_KEYCODE) {
            toggleDrawer();
        }
    }, [isDrawerOpen, toggleDrawer]);

    useEffect(() => {
        document.body.addEventListener('keydown', handleKeyDown);

        return () => {
            document.body.removeEventListener('keydown', handleKeyDown);
        };
    }, [handleKeyDown]);

    return (
        <div
            role="none"
            className={ backdropClassName }
            onClick={ toggleDrawer } />
    );
};

Backdrop.propTypes = {
    className: PropTypes.string,
    isDrawerOpen: PropTypes.bool.isRequired,
    toggleDrawer: PropTypes.func.isRequired,
};

export default Backdrop;
