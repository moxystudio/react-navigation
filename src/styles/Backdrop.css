@import "./variables.css";

.react-navigation__backdrop {
    position: absolute;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    z-index: var(--z-index-base);
    transform: translateX(-100%);
    opacity: 0;
    background-color: color(var(--color-rose-gold) alpha(0.8));
    transition: opacity var(--duration-standard);

    &.show {
        transform: translateX(0);
        opacity: 0.1;
    }
}
