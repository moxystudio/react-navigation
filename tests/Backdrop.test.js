import React from 'react';
import { render, screen } from '@testing-library/react';

import Backdrop from '../src/components/Backdrop';
import AppTree from './AppTree';

const renderWithProps = (props = {}, Component = Backdrop) => render(
    <AppTree>
        <Component { ...props } />
    </AppTree>,
);

describe('Backdrop', () => {
    it('should render correctly', () => {
        const { asFragment } = renderWithProps();

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly with className', () => {
        renderWithProps({ className: 'foo' });

        expect(screen.getByTestId('backdrop')).toHaveClass('foo');
    });

    it('should be invisible when the Drawer is closed', () => {
        renderWithProps();

        expect(screen.getByTestId('backdrop')).not.toHaveClass('show');
    });

    it('should be visible when the Drawer is open', () => {
        const mockNavigationContextValue = {
            isDrawerOpen: true,
            toggleDrawer: jest.fn(),
        };

        render(
            <AppTree navigationContextValue={ mockNavigationContextValue }>
                <Backdrop />
            </AppTree>,
        );

        expect(screen.getByTestId('backdrop')).toHaveClass('show');
    });
});

