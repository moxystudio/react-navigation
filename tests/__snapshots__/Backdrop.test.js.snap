// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop should render correctly 1`] = `
<DocumentFragment>
  <div
    class="react-navigation__backdrop"
    data-testid="backdrop"
    role="none"
  />
</DocumentFragment>
`;
